pandas/core/indexes/base.py

```diff
@@ -6521,6 +6521,31 @@ def drop(
             indexer = indexer[~mask]
         return self.delete(indexer)
 
+    def infer_objects(self, copy: bool = True) -> Index:
+        """
+        If we have an object dtype, try to infer a non-object dtype.
+
+        Parameters
+        ----------
+        copy : bool, default True
+            Whether to make a copy in cases where no inference occurs.
+        """
+        if self.dtype != object or self._is_multi:
```

Comment:
Should MultiIndex just raise a `NotImplementedError` instead and just suggest to use `to_frame().infer_objects()`?

Reply:
good idea, will update.